osquery/tables/system/linux/services.cpp

```diff
@@ -0,0 +1,774 @@
+                                   job_type,
+                                   job_path,
+                                   0U};
+
+      auto status = getSystemdUnitPid(unit_info.process_id, unit_path, bus);
+      if (!status.ok()) {
+        VLOG(1) << "Failed to determine the process id for the following "
+                   "systemd unit: "
+                << unit_path << ". Error: " << status.getMessage();
+      }
+
+      unit_list.push_back(unit_info);
+    }
+
+    if (sd_bus_message_exit_container(reply) < 0) {
+      throw std::runtime_error("Failed to exit the data container");
+    }
+
+    reply = sd_bus_message_unref(reply);
+    message = sd_bus_message_unref(message);
+
+    return Status(0, "OK");
+
+  } catch (const std::exception& e) {
+    if (reply != nullptr) {
+      reply = sd_bus_message_unref(reply);
+    }
+
+    if (message != nullptr) {
+      message = sd_bus_message_unref(message);
+    }
+
+    return Status(1, e.what());
+  }
+}
+
+Status getInactiveSystemdUnitList(std::vector<SystemdUnitInfo>& unit_list,
+                                  sd_bus* bus) {
+  unit_list.clear();
+
+  sd_bus_message* message = nullptr;
+  sd_bus_message* reply = nullptr;
+  sd_bus_error bus_error = SD_BUS_ERROR_NULL;
+
+  try {
+    if (sd_bus_call_method(bus,
+                           "org.freedesktop.systemd1",
+                           "/org/freedesktop/systemd1",
+                           "org.freedesktop.systemd1.Manager",
+                           "ListUnitFiles",
+                           &bus_error,
+                           &reply,
+                           nullptr) < 0) {
+      std::string error_message;
+      if (bus_error.message != nullptr) {
+        error_message = bus_error.message;
+      } else {
+        error_message = "Failed to call the remote method";
+      }
+
+      throw std::runtime_error(error_message);
+    }
+
+    if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "(ss)") < 0) {
+      throw std::runtime_error("Failed to enter the data container");
+    }
+
+    while (true) {
+      const char* unit_path = nullptr;
+      const char* unit_state = nullptr;
+
+      auto error = sd_bus_message_read(reply, "(ss)", &unit_path, &unit_state);
+      if (error == 0) {
+        break;
+      } else if (error < 0) {
+        throw std::runtime_error("Failed to parse the unit information");
+      }
+
+      if (std::strcmp(unit_state, "disabled") != 0) {
+        continue;
+      }
+
+      SystemdUnitInfo info = {};
+      info.unit_path = unit_path;
+      info.active_state = unit_state;
+
+      unit_list.push_back(info);
+    }
+
+    if (sd_bus_message_exit_container(reply) < 0) {
+      throw std::runtime_error("Failed to exit the data container");
+    }
+
+    reply = sd_bus_message_unref(reply);
+    message = sd_bus_message_unref(message);
+
+    if (unit_list.empty()) {
+      return Status(1, "No services returned by the manager!");
```

Comment:
Is it an error if no inactive service exists?

Reply:
Yes, as there are many important services that are supposed to always exists (like the shutdown handler)